tracer/src/Datadog.Trace/Tracer.cs

```diff
@@ -336,19 +336,15 @@ internal SpanContext CreateSpanContext(ISpanContext parent = null, string servic
                 parent = DistributedTracer.Instance.GetSpanContext() ?? TracerManager.ScopeManager.Active?.Span?.Context;
             }
 
-            TraceContext traceContext;
+            // get trace context from parent...
+            var parentSpanContext = parent as SpanContext;
+            var traceContext = parentSpanContext?.TraceContext;
 
-            // try to get the trace context (from local spans) or
-            // sampling priority (from propagated spans),
-            // otherwise start a new trace context
-            if (parent is SpanContext parentSpanContext)
+            if (traceContext == null)
             {
-                traceContext = parentSpanContext.TraceContext
-                    ?? new TraceContext(this) { SamplingPriority = parentSpanContext.SamplingPriority ?? DistributedTracer.Instance.GetSamplingPriority() };
-            }
-            else
-            {
-                traceContext = new TraceContext(this) { SamplingPriority = DistributedTracer.Instance.GetSamplingPriority() };
+                // ...or start a new one
+                traceContext = new TraceContext(this);
+                traceContext.SetSamplingPriority(parentSpanContext?.SamplingPriority ?? DistributedTracer.Instance.GetSamplingPriority());
```

Comment:
If I did this correctly, there should be no change in behavior. The logic to try to get a `TraceContext` is something like:
- if the parent is a `SpanContext` and it has a `TraceContext`, use that one (and don't change its sampling priority)
- otherwise, create a new `TraceContext`, and set its sampling priority to:
  - if the parent is a `SpanContext` and has a `SamplingPriority`, use that one
  - otherwise, try `DistributedTracer.Instance.GetSamplingPriority()`
  - otherwise, leave it `null`

What I changed was move the initial `traceContext = (parent as SpanContext)?.TraceContext` up and out of the `if` (see lines 340-341 above), and then handle the case where that was null. I think it's easier to understand the logic now rather than having two separate `new TraceContext(this)` calls and two different expressions for `SamplingPriority = ...`.

Reply:
Also note that either `SpanContext.TraceContext` has a value (local spans) or `SpanContext.SamplingPriority` has a value (propagated upstream span), never both.

Reply:
Yeah, looks like that's still correct then. Tracing through all the null propagations to check this got me confused 😅 

It bugs me a bit that the "logic" flow is completely different to the code flow though... I wonder if making the code more "dumb" but expressing that might be beneficial e.g. extracting to a local function:

```csharp
private static GetTraceContext(ISpanContext parent)
{
    if (parent is SpanContext parentSpanContext && parentSpanContext.TraceContext is {} parentTraceContext)
    {
        return parentTraceContext;
    }

    var traceContext = new TraceContext(this);
    if (parent is SpanContext parentSpanContext & parentSpanContext.SamplingPriority is {} parentPriority)
    {
        traceContext.SetSamplingPriority(parentPriority);
    } 
    else if (DistributedTracer.Instance.GetSamplingPriority() is {} distributedPriority)
    {
        traceContext.SetSamplingPriority(distributedPriority);
    }
    
    return traceContext;
}
```

It's more lines, but it follows your logic at least?

Reply:
I'm gonna revert this change then 😅. It was unrelated to this PR (removing `TraceContext.SamplingPriority` setter). I just happened to see this while I was making that change, and it always bugged me.

`Tracer.CreateSpanContext()` is going away in my refactoring PR anyway 🤞🏽 Thanks!

Reply:
Reverted in 1bdda3ddebb2e76f03bc1b03efa416239c459b62.

Reply:
To be fair, I don't that your new code was any "worse" than the previous code! IMO, neither one makes the logic you describe entirely clear 😅

Reply:
Yeah I got it, no worries! I just prefer not to change it unless is clearly better. Especially if I'm trying to sneak it into a PR for something else.